src/Workspaces/Core/Portable/Utilities/NullableHelpers/TypeSymbolWithNullableAnnotation.cs

```diff
@@ -25,11 +25,21 @@ protected TypeSymbolWithNullableAnnotation(ITypeSymbol wrappedSymbol, NullableAn
             }
 
             public bool Equals(ISymbol other)
+            {
+                var t2 = other as ITypeSymbol;
```

Comment:
i'm not really understanding how the IDE side works.  ITypeComparable is internal, and so are all the TypeCompareKinds... so how does the IDE call in and ask for specific types of comparing?  For example, the IDE already had to write a whole SymbolEquivalenceComparer to try to deal with the compiler not exposing these concepts.  it seems like we're still there.

It would be really good if this system were sufficient and flexible enough so that we could remove all (or most) of the code in SymbolEquivalenceComparer and replace with SymbolEqualityComparer.

Reply:
The current goal and exposed options are mainly to preserve existing analyzer semantics and allow authors to opt into nullability in comparisons as appropriate. We can potentially expose more symbol equality comparers in the future with different options, but the idea is that you'd use one of the predefined SymbolEqualityComparer instances.

---
In reply to: [303751120](https://github.com/dotnet/roslyn/pull/37247#discussion_r303751120) [](ancestors = 303751120)